File: src/main/java/com/blackwing/easyExploration/EasyExploration.java
package com.blackwing.easyExploration;

import com.blackwing.easyExploration.handlers.ConfigurationHandler;
import com.blackwing.easyExploration.handlers.SaveInventoryHandler;
import com.blackwing.easyExploration.handlers.ShowDamageHandler;
import com.blackwing.easyExploration.handlers.ShowDeathLocationHandler;
import com.blackwing.easyExploration.util.EventHandlerBase;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLFingerprintViolationEvent;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@Mod(
        useMetadata = true,
        modid = EasyExploration.MODID,
        name = EasyExploration.NAME,
        version = EasyExploration.VERSION,
        acceptedMinecraftVersions = "1.12",
        canBeDeactivated = true,
        certificateFingerprint = EasyExploration.FINGERPRINT
)
public class EasyExploration extends EventHandlerBase {
    public static final String MODID = "easyexploration";
    public static final String NAME = "Easy Exploration";
    public static final String VERSION = "@VERSION@";
    public static final String FINGERPRINT = "@FINGERPRINT@";

    @Mod.Instance(MODID)
    public static EasyExploration instance;

    protected final Logger log = LogManager.getLogger(EasyExploration.MODID + "." + getClass().getSimpleName());

    /*
     * Here we check if our mod has been tampered with
     */
    @Mod.EventHandler
    public void invalidFingerprint(FMLFingerprintViolationEvent event) {
        if (!FINGERPRINT.equals("@FINGERPRINT@")) {
            log.warn("THE EASYEXPLORATION MOD THAT YOU ARE USING HAS BEEN CHANGED/TAMPERED WITH!");
        }
    }

    // public static final SimpleNetworkWrapper packetHandler = NetworkRegistry.instance.newSimpleChannel(MODID);

    public static final ConfigurationHandler configHandler = new ConfigurationHandler();
    public static final SaveInventoryHandler saveInventoryHandler = new SaveInventoryHandler();
    public static final ShowDeathLocationHandler showDeathLocationHandler = new ShowDeathLocationHandler();
    public static final ShowDamageHandler showDamageHandler = new ShowDamageHandler();

    protected static final EventHandlerBase[] handlers = {
            configHandler,
            saveInventoryHandler,
            showDeathLocationHandler,
            showDamageHandler
    };

    /*
     * This is where we load our network configuration, mod configuration, .. and where we initialize our items and blocks
     */
    @Mod.EventHandler
    public void onPreInit(FMLPreInitializationEvent event) {
        // register event handlers
        super.onPreInit(event);
        for (EventHandlerBase handler : handlers) handler.onPreInit(event);
        // register Messages
    }

    /*
     * This is where we register our GUIs, tile entities, crafting recipes, ..
     */
    @Mod.EventHandler
    public void onInit(FMLInitializationEvent event) {
        super.onInit(event);
        for (EventHandlerBase handler : handlers) handler.onInit(event);
    }

    /*
     * This is where you can run things after all other mods have e.g. registered their blocks and items
     */
    @Mod.EventHandler
    public void onPostInit(FMLPostInitializationEvent event) {
        super.onPostInit(event);
        for (EventHandlerBase handler : handlers) handler.onPostInit(event);
    }
}
//TODO: FIX [21:22:54] [main/ERROR] [FML]: FML appears to be missing any signature data. This is not a good thing
// TODO: create an actual fingerprint

File: src/main/scala/com/blackwing/easyExploration/EasyExploration.scala
package com.blackwing.easyExploration

import com.blackwing.easyExploration.proxy.CommonProxy
import com.blackwing.easyExploration.utilities.{EventHandlerBase, IHasModel}
import net.minecraft.block.Block
import net.minecraft.item.Item
import net.minecraftforge.client.event.ModelRegistryEvent
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.common.event.{FMLFingerprintViolationEvent, FMLInitializationEvent, FMLPostInitializationEvent, FMLPreInitializationEvent}
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.{Mod, SidedProxy}

@Mod(
	useMetadata = true,
	modid = EasyExploration.modId,
	name = EasyExploration.name,
	version = EasyExploration.version,
	acceptedMinecraftVersions = "1.12",
	canBeDeactivated = true,
	certificateFingerprint = EasyExploration.fingerprint,
	modLanguage = "scala"
)
object EasyExploration extends EventHandlerBase {

	val modId: String = Reference.modId
	val name: String = Reference.name
	val version: String = Reference.version
	val fingerprint: String = Reference.fingerprint

	@SidedProxy(clientSide = "com.blackwing.easyExploration.proxy.ClientProxy", serverSide = "com.blackwing.easyExploration.proxy.CommonProxy")
	val proxy: CommonProxy.type = _

	/*
	 * Here we check if our mod has been tampered with
	 */
	@Mod.EventHandler
	def invalidFingerprint(event: FMLFingerprintViolationEvent): Unit =
		if (fingerprint != "@FINGERPRINT@") logger.warn("THE EASYEXPLORATION MOD THAT YOU ARE USING HAS BEEN CHANGED/TAMPERED WITH!")

	// public static final SimpleNetworkWrapper packetHandler = NetworkRegistry.instance.newSimpleChannel(MODID)

	@Mod.EventHandler
	override def onPreInit(event: FMLPreInitializationEvent): Unit = {
		super.onPreInit(event)
		proxy.onPreInit(event)
	}

	@Mod.EventHandler
	override def onInit(event: FMLInitializationEvent): Unit = {
		super.onInit(event)
		proxy.onInit(event)
	}

	@Mod.EventHandler
	override def onPostInit(event: FMLPostInitializationEvent): Unit = {
		super.onPostInit(event)
		proxy.onPostInit(event)
	}

	var items: List[Item] = List.empty
	var blocks: List[Block] = List.empty

	@SubscribeEvent
	def onItemRegister(event: RegistryEvent.Register[Item]): Unit =
		event.getRegistry.registerAll(items: _*)

	@SubscribeEvent
	def onBlockRegister(event: RegistryEvent.Register[Block]): Unit =
		event.getRegistry.registerAll(blocks: _*)

	@SubscribeEvent
	def onModelRegister(event: ModelRegistryEvent): Unit = {
		items.filter(_.isInstanceOf[IHasModel]).foreach(_.asInstanceOf[IHasModel].registerModels())

		blocks.filter(_.isInstanceOf[IHasModel]).foreach(_.asInstanceOf[IHasModel].registerModels())
	}
}
